File: implementations/python/src/esque/elasticsearch/documents.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any

MIGRATION_INDEX = ".esque"
LOCK_ID_PREFIX = "lock"

INDEX_DEFINITION: dict[str, Any] = {
    "settings": {
        "index": {
            "number_of_shards": "1",
            "auto_expand_replicas": "0-all",
            "refresh_interval": "1s",
        }
    },
    "mappings": {
        "properties": {
            "lock": {"properties": {"date": {"type": "date"}}},
            "migration": {
                "properties": {
                    "checksum": {"type": "long"},
                    "description": {"type": "keyword"},
                    "executionTime": {"type": "long"},
                    "filename": {"type": "keyword"},
                    "installedOn": {"type": "date"},
                    "migrationKey": {"type": "keyword"},
                    "order": {"type": "long"},
                    "version": {"type": "keyword"},
                }
            },
        }
    },
}


@dataclass
class MigrationRecord:
    migration_key: str
    order: int
    filename: str
    version: str
    description: str
    checksum: int
    installed_by: str | None
    installed_on: datetime
    execution_time: int

    def to_document(self) -> dict[str, Any]:
        doc: dict[str, Any] = {
            "migrationKey": self.migration_key,
            "order": self.order,
            "filename": self.filename,
            "version": self.version,
            "description": self.description,
            "checksum": self.checksum,
            "installedOn": self.installed_on.isoformat(),
            "executionTime": self.execution_time,
        }
        if self.installed_by is not None:
            doc["installedBy"] = self.installed_by
        return {"migration": doc}

    @classmethod
    def from_document(cls, source: dict[str, Any]) -> MigrationRecord:
        raw: dict[str, Any] = source["migration"]
        return cls(
            migration_key=str(raw["migrationKey"]),
            order=int(raw["order"]),
            filename=str(raw["filename"]),
            version=str(raw["version"]),
            description=str(raw["description"]),
            checksum=int(raw["checksum"]),
            installed_by=str(raw["installedBy"]) if raw.get("installedBy") is not None else None,
            installed_on=datetime.fromisoformat(str(raw["installedOn"])),
            execution_time=int(raw["executionTime"]),
        )


@dataclass
class MigrationLock:
    date: datetime

    def to_document(self) -> dict[str, Any]:
        return {"lock": {"date": self.date.isoformat()}}

File: implementations/typescript/src/elasticsearch/documents.ts
export const MIGRATION_INDEX = ".esque";
export const LOCK_ID_PREFIX = "lock";

export const INDEX_DEFINITION = {
  settings: {
    index: {
      number_of_shards: "1",
      auto_expand_replicas: "0-all",
      refresh_interval: "1s",
    },
  },
  mappings: {
    properties: {
      lock: { properties: { date: { type: "date" } } },
      migration: {
        properties: {
          checksum: { type: "long" },
          description: { type: "keyword" },
          executionTime: { type: "long" },
          filename: { type: "keyword" },
          installedOn: { type: "date" },
          migrationKey: { type: "keyword" },
          order: { type: "long" },
          version: { type: "keyword" },
        },
      },
    },
  },
} as const;

export interface MigrationRecord {
  readonly migrationKey: string;
  readonly order: number;
  readonly filename: string;
  readonly version: string;
  readonly description: string;
  readonly checksum: number;
  readonly installedBy: string | null;
  readonly installedOn: string; // ISO-8601 UTC
  readonly executionTime: number;
}

// Wrapper-object serialization — mirrors JVM @JsonTypeInfo(As.WRAPPER_OBJECT).
export function migrationRecordToDocument(record: MigrationRecord): { migration: Record<string, unknown> } {
  const doc: Record<string, unknown> = {
    migrationKey: record.migrationKey,
    order: record.order,
    filename: record.filename,
    version: record.version,
    description: record.description,
    checksum: record.checksum,
    installedOn: record.installedOn,
    executionTime: record.executionTime,
  };
  if (record.installedBy !== null) {
    doc.installedBy = record.installedBy;
  }
  return { migration: doc };
}

export function migrationRecordFromDocument(source: Record<string, unknown>): MigrationRecord {
  const raw = source.migration as Record<string, unknown>;
  return {
    migrationKey: String(raw.migrationKey),
    order: Number(raw.order),
    filename: String(raw.filename),
    version: String(raw.version),
    description: String(raw.description),
    checksum: Number(raw.checksum),
    installedBy: raw.installedBy !== undefined && raw.installedBy !== null ? String(raw.installedBy) : null,
    installedOn: String(raw.installedOn),
    executionTime: Number(raw.executionTime),
  };
}

export function migrationLockToDocument(date: Date): { lock: { date: string } } {
  return { lock: { date: date.toISOString() } };
}
